Add KOKKOS versions of fixes

alphataubio opened the pull request:
**Summary**

<!--Briefly describe the new feature(s), enhancement(s), or bugfix(es) included in this pull request.-->

Add KOKKOS versions:

- [x] `fix cmap/kk` (to have all needed fixes in KOKKOS for performance on GPUs)

- [x] `fix nve/limit/kk` (to relax overlapped systems with KOKKOS)

- [x] `fix recenter/kk` (to have all needed fixes in KOKKOS for performance on GPUs), thanks to @srtee for suggesting this fix on matsci, very useful.

- [x] `fix wall/region/kk` (to have all needed fixes in KOKKOS for performance on GPUs)

- [x] `region sphere/kk` (bugfix for dynamic_cast in FixWallRegionKokkos)

- [x] `GroupKokkos` (device functions for mass / xcm / vcm / angmom / inertia)

- [x] `fix momentum/kk` (backport GroupKokkos changes from fix recenter/kk)


**Related Issue(s)**

n/a

**Author(s)**

<!--Please state name and affiliation of the author or authors that should be credited with the changes in this pull request. If this pull request adds new files to the distribution, please also provide a suitable "long-lived" e-mail address (ideally something that can outlive your institution's e-mail, in case you change jobs) for the *corresponding* author, i.e. the person the LAMMPS developers can contact directly with questions and requests related to maintenance and support of this contributed code.-->

@alphataubio 

**Licensing**

By submitting this pull request, I agree, that my contribution will be included in LAMMPS and redistributed under either the GNU General Public License version 2 (GPL v2) or the GNU Lesser General Public License version 2.1 (LGPL v2.1).

**Backward Compatibility**

n/a

**Implementation Notes**

<!--Provide any relevant details about how the changes are implemented, how correctness was verified, how other features - if any - in LAMMPS are affected-->

ADDED UNIT TESTS:
- fix-timestep-recenter-coords.yaml
- fix-timestep-recenter-init.yaml
- fix-timestep-recenter-null.yaml
- fix-timestep-wall_region_lj1043.yaml
- fix-timestep-wall_region_lj126.yaml
- fix-timestep-wall_region_lj93.yaml
- fix-timestep-wall_region_morse.yaml


**Post Submission Checklist**

<!--Please check the fields below as they are completed **after** the pull request has been submitted. Delete lines that don't apply-->

- [x] The feature or features in this pull request is complete
- [x] Licensing information is complete
- [x] Corresponding author information is complete
- [x] The source code follows the LAMMPS formatting guidelines
- [x] Suitable new documentation files and/or updates to the existing docs are included
- [x] The added/updated documentation is integrated and tested with the documentation build system
- [x] The feature has been verified to work with the conventional build system
- [x] The feature has been verified to work with the CMake based build system
- [x] Suitable tests have been added to the unittest tree.
- [ ] A package specific README file has been included or updated
- [x] One or more example input decks are included

**Further Information, Files, and Links**

<!--Put any additional information here, attach relevant text or image files, and URLs to external sites (e.g. DOIs or webpages)-->

alphataubio committed:
added dihedral charmfsw kokkos test

alphataubio committed:
Revert "added dihedral charmfsw kokkos test"

This reverts commit 55b1b5d008c54cc2400eee80e603b307bb7bfb1c.

alphataubio committed:
Merge branch 'lammps:develop' into alphataubio-kokkos-bio

alphataubio committed:
first draft fix recenter/kk and unit test

alphataubio committed:
fix recenter/kk works with coords, init, null plus 3 tests

alphataubio committed:
first draft fix cmap/kk

alphataubio committed:
add cmap unit test

alphataubio committed:
add more crossterms to cmap unit test

alphataubio committed:
make operator and inline functions const, plus remove printf debugging

alphataubio committed:
Merge branch 'lammps:develop' into alphataubio-kokkos-bio

alphataubio committed:
Merge branch 'alphataubio-kokkos-bio' of https://github.com/alphataubio/lammps-alphataubio into alphataubio-kokkos-bio

alphataubio committed:
add fix wall/region/kk

alphataubio committed:
add 4 units tests for fix wall/region/kk

alphataubio committed:
author information only in cpp, kokkos version needs protected instead of private variables

alphataubio committed:
cleanup

alphataubio committed:
added fix nve/limit/kk

alphataubio committed:
bugfix for compute_scalar() to work, test PASSED

alphataubio committed:
add charge to atom style template

alphataubio committed:
fix typo in comment

alphataubio committed:
fix molecule add for MESH and remove 'not compatible' error message

alphataubio committed:
update atom_style template test for added charge

alphataubio committed:
added RegSphereKokkos to bugfix dynamic_cast in FixWallRegionKokkos

alphataubio committed:
oops

alphataubio committed:
update email and fix typo

alphataubio committed:
update region doc for sphere/kk

alphataubio committed:
first draft charmm_c36_jul24.gz to LAMMPS data converter

alphataubio committed:
added pair coeffs

alphataubio committed:
Merge branch 'lammps:develop' into alphataubio-kokkos-fixes

alphataubio committed:
add 3d float views to ArrayTypes<LMPHostType>

alphataubio committed:
Merge branch 'alphataubio-kokkos-fixes' of https://github.com/alphataubio/lammps-alphataubio into alphataubio-kokkos-fixes

alphataubio committed:
KOKKOS_INLINE_FUNCTION powint() to replace MathSpecial::powint

alphataubio committed:
fix cuda warnings

alphataubio committed:
fix cuda warnings

alphataubio committed:
auto atomKK variables for lambda capture

alphataubio committed:
Update fix_nve_limit_kokkos.cpp

alphataubio committed:
auto var lambda capture

alphataubio committed:
fix warnings

alphataubio committed:
kokkos atan2

alphataubio committed:
cleanup

alphataubio committed:
Update fix_recenter_kokkos.cpp

alphataubio committed:
fix warnings

alphataubio committed:
match base class in kokkos

alphataubio committed:
Revert "match base class in kokkos"

This reverts commit 36494af3c73463411c3e46a7a496ce3bd6707bfb.

alphataubio committed:
match base class in kokkos

alphataubio committed:
fix warnings

alphataubio committed:
fix warnings

alphataubio committed:
cleanup

alphataubio committed:
bugfix

alphataubio committed:
oops

alphataubio committed:
Update fix_nve_limit_kokkos.cpp

alphataubio committed:
Merge branch 'lammps:develop' into alphataubio-kokkos-fixes

alphataubio committed:
kokkos parallel_for

alphataubio committed:
Merge branch 'alphataubio-kokkos-fixes' of https://github.com/alphataubio/lammps-alphataubio into alphataubio-kokkos-fixes

alphataubio committed:
revert to original cmap plus minimal changes for kk

alphataubio committed:
replace error->all() with kokkos::abort()

alphataubio committed:
feeble attempt

alphataubio committed:
Update fix_cmap_kokkos.cpp

alphataubio committed:
parallel_scan

alphataubio committed:
thanks @crtrott

alphataubio committed:
fix warnings

alphataubio committed:
KOKKOS_CLASS_LAMBDA

alphataubio committed:
tagged operators

alphataubio committed:
cleanup

alphataubio committed:
explicit warning(s) which fix sort_device=0

alphataubio committed:
explicit warning(s) exchange_comm_device=0

alphataubio committed:
sort on device

alphataubio committed:
oops

alphataubio committed:
pack/unpack on device

alphataubio committed:
replace ubuf with static_cast

alphataubio committed:
auto variables for lambda capture

alphataubio committed:
oops

alphataubio committed:
update rst kk

alphataubio committed:
first draft delete_atoms/kk

alphataubio committed:
first draft

alphataubio committed:
resolve conflict

alphataubio committed:
Merge branch 'lammps:develop' into alphataubio-kokkos-fixes

alphataubio committed:
punt to later

alphataubio committed:
Merge branch 'alphataubio-kokkos-fixes' of https://github.com/alphataubio/lammps-alphataubio into alphataubio-kokkos-fixes

alphataubio committed:
remove whitespace

alphataubio committed:
punt again

alphataubio committed:
punt drude too

alphataubio committed:
compute_vector() bugfix

alphataubio committed:
regenerate fix recenter tests

alphataubio committed:
regenerate fix wall/region tests

stanmoore1 commented:
Running this PR through internal Kokkos regression testing now.

alphataubio committed:
Update fix_nve_limit.rst

alphataubio committed:
Update fix_recenter.rst

alphataubio committed:
Update fix_wall_region.rst

alphataubio committed:
Update region.rst

alphataubio committed:
Update Commands_fix.rst

alphataubio committed:
Update fix_cmap.rst

alphataubio committed:
update kk index entries

akohlmey commented on src/REACTION/fix_bond_react.cpp:
```diff
@@ -611,6 +611,8 @@ FixBondReact::FixBondReact(LAMMPS *lmp, int narg, char **arg) :
 
 FixBondReact::~FixBondReact()
 {
+  if (copymode) return; // needed for KOKKOS [alphataubio,2024/08]
```
Since there is no fix bond/react/kk here, these changes should be reverted.

alphataubio replied:
@stanmoore1 asked me to wrap up this PR if possible with what i have so far  because he needed `recenter/kk` for his own work. i can revert this commit, stash it in the branch, and resubmit another PR from this branch later with `bond/react/kk`

alphataubio replied:
done

akohlmey commented on src/REACTION/fix_bond_react.h:
```diff
@@ -55,7 +51,7 @@ class FixBondReact : public Fix {
   double memory_usage() override;
 
- private:
+ protected:
```
Since there is no fix bond/react/kk here, these changes should be reverted.

alphataubio replied:
done

akohlmey commented on src/REACTION/fix_bond_react.h:
```diff
@@ -11,10 +11,6 @@
    See the README file in the top-level LAMMPS directory.
 ------------------------------------------------------------------------- */
 
-/* ----------------------------------------------------------------------
```
Please never remove any attribution.

alphataubio replied:
done

akohlmey commented on src/create_atoms.cpp:
```diff
@@ -303,7 +303,7 @@ void CreateAtoms::command(int narg, char **arg)
     if (onemol->typeflag == 0) error->all(FLERR, "Create_atoms molecule must have atom types");
     if (ntype + onemol->ntypes <= 0 || ntype + onemol->ntypes > atom->ntypes)
-      error->all(FLERR, "Invalid atom type in create_atoms mol command");
+      error->all(FLERR, "Invalid atom type {} in create_atoms mol command (onemol->ntypes {} atom->ntypes {})", ntype, onemol->ntypes, atom->ntypes);
```
```suggestion
      error->all(FLERR, "Invalid atom type {} in create_atoms mol command", ntype + onemol->ntypes);
```
For molecules, ntype is only the offset added to the atom types in the molecule file.

akohlmey commented on src/create_atoms.cpp:
```diff
@@ -319,11 +319,7 @@ void CreateAtoms::command(int narg, char **arg)
-      error->all(FLERR, "Create_atoms mesh is not compatible with the 'mol' option");
-    if (scaleflag) error->all(FLERR, "Create_atoms mesh must use 'units box' option");
-  }
+  if (style == MESH && scaleflag) error->all(FLERR, "Create_atoms mesh must use 'units box' option");
```
Using the "mesh" option with molecule files makes no sense. These changes need to be reverted.

alphataubio replied:
see my comment on slack.

alphataubio replied:
done. i copied my changed file outside local repo, downloaded file from develop branch, committed that, moved my modified file back in local repo and stashed that

akohlmey commented on src/create_atoms.cpp:
```diff
@@ -968,7 +964,12 @@ int CreateAtoms::add_bisection(const double vert[3][3], tagint molid)
         (center[1] < subhi[1]) && (center[2] >= sublo[2]) && (center[2] < subhi[2])) {
 
-      atom->avec->create_atom(ntype, center);
+      if (mode == ATOM) atom->avec->create_atom(ntype, center);
```
Using the "mesh" option with molecule files makes no sense. These changes need to be reverted.

akohlmey commented on src/create_atoms.cpp:
```diff
@@ -1050,7 +1051,12 @@ int CreateAtoms::add_quasirandom(const double vert[3][3], tagint molid)
         (point[1] < subhi[1]) && (point[2] >= sublo[2]) && (point[2] < subhi[2])) {
 
-      atom->avec->create_atom(ntype, point);
+      if (mode == ATOM) atom->avec->create_atom(ntype, point);
```
Using the "mesh" option with molecule files makes no sense. These changes need to be reverted.

akohlmey commented on src/fix_recenter.cpp:
```diff
@@ -217,7 +217,11 @@ void FixRecenter::initial_integrate_respa(int vflag, int ilevel, int /*iloop*/)
+  //if (ilevel == nlevels_respa-1) initial_integrate(vflag);
+  // FIXME: why does always calling initial_integrate make respa tests
+  // pass, i dont know !
+  initial_integrate(vflag);
```
The force style unit tests use r-RESPA under the very special condition of having a multiplier of 1 between 2 different RESPA levels so the resulting time integration would be equivalent to verlocity verlet.

To make the test work *and* produce correct results with the fix under general conditions, some more work and more complex testing with other r-RESPA setups is needed.

This change needs to be reverted.

akohlmey commented on src/region_sphere.h:
```diff
@@ -36,7 +36,7 @@ class RegSphere : public Region {
   void velocity_contact_shape(double *, double *) override;
 
- private:
+ protected: // KOKKOS subclass needs protected not private
```
```suggestion
 protected:
```

akohlmey commented on tools/charmm36/charmm36.py:
```diff
@@ -0,0 +1,201 @@
+
```
Added tools must be properly described in the manual and have a README file in their folder explaining its specific use and a one line description added to tools/README.

This one doesn't look like it is ready for general use but rather specific for your personal use. I assume its commit is accidental and thus it should be removed.

alphataubio replied:
done

akohlmey commented on src/MOLECULE/atom_vec_template.cpp:
```diff
@@ -29,20 +29,21 @@ AtomVecTemplate::AtomVecTemplate(LAMMPS *lmp) : AtomVec(lmp)
   atom->molecule_flag = 1;
   atom->molindex_flag = 1;
   atom->molatom_flag = 1;
+  atom->q_flag = 1;
```
This change must be reverted.

Atom style template is equivalent to atom style molecular. 
The equivalent of atom style full is achieved through using atom style hybrid.

alphataubio replied:
done

akohlmey commented on unittest/formats/test_atom_styles.cpp:
```diff
@@ -2743,6 +2743,7 @@ TEST_F(AtomStyleTest, template)
     expected.has_x         = true;
     expected.has_v         = true;
     expected.has_f         = true;
+    expected.q_flag        = 1;
```
This change must be reverted. See my comment in the corresponding AtomVec class

alphataubio replied:
done

akohlmey commented on unittest/formats/test_atom_styles.cpp:
```diff
@@ -3013,437 +3014,9 @@ TEST_F(AtomStyleTest, template)
     ASSERT_EQ(lmp->atom->nghost, 0);
     ASSERT_NE(lmp->atom->nmax, -1);
     ASSERT_EQ(lmp->atom->tag_enable, 1);
-    ASSERT_EQ(lmp->atom->q_flag, 0);
```
These tests must be re-instated. see my previous comments.

alphataubio replied:
done

akohlmey reviewed (changes requested):
There are several changes that must be reverted before this pull request can be merged. Please see my individual comments.

alphataubio committed:
Update src/create_atoms.cpp

Co-authored-by: Axel Kohlmeyer <akohlmey@gmail.com>

alphataubio committed:
Update src/region_sphere.h

Co-authored-by: Axel Kohlmeyer <akohlmey@gmail.com>

alphataubio committed:
Revert "update atom_style template test for added charge"

This reverts commit 3ea74b17252ef4a4e16d2ee6841c518e983e435a.

alphataubio committed:
revert changes requested by @akohlmey

alphataubio committed:
Delete charmm36.py

alphataubio committed:
Update atom_vec_template.cpp

alphataubio committed:
Update test_atom_styles.cpp

alphataubio committed:
Update create_atoms.cpp

akohlmey commented on unittest/formats/test_atom_styles.cpp:
```diff
@@ -347,7 +347,6 @@ void ASSERT_ATOM_STATE_EQ(Atom *atom, const AtomState &expected)
     ASSERT_ARRAY_ALLOCATED(atom->x, expected.has_x);
     ASSERT_ARRAY_ALLOCATED(atom->v, expected.has_v);
     ASSERT_ARRAY_ALLOCATED(atom->f, expected.has_f);
-    ASSERT_ARRAY_ALLOCATED(atom->q, expected.q_flag);
```
Please restore this line.

alphataubio replied:
done

akohlmey commented on unittest/force-styles/tests/fix-timestep-wall_region_morse.yaml:
```diff
@@ -0,0 +1,84 @@
+---
+lammps_version: 29 Aug 2024
+tags: generated
```
The "generated" needs to be removed before adding or updating yaml files.
The presence of the tag is used an indicator that a test has been updated but not yet manually reviewed.

alphataubio replied:
done

alphataubio replied:
done

akohlmey commented on unittest/force-styles/tests/fix-timestep-wall_region_lj1043.yaml:
```diff
@@ -0,0 +1,84 @@
+---
+lammps_version: 29 Aug 2024
+tags: generated
```
Another generated tag that needs to be removed

alphataubio replied:
done

akohlmey commented on unittest/force-styles/tests/fix-timestep-wall_region_lj126.yaml:
```diff
@@ -0,0 +1,84 @@
+---
+lammps_version: 29 Aug 2024
+tags: generated
```
Another generated tag that needs to be removed

alphataubio replied:
done

akohlmey commented on unittest/force-styles/tests/fix-timestep-wall_region_lj93.yaml:
```diff
@@ -0,0 +1,84 @@
+---
+lammps_version: 29 Aug 2024
+tags: generated
```
Another generated tag that needs to be removed

alphataubio replied:
done

akohlmey commented on unittest/force-styles/tests/fix-timestep-wall_region_harmonic.yaml:
```diff
@@ -1,6 +1,7 @@
-lammps_version: 3 Aug 2022
-date_generated: Mon Aug 15 01:14:02 2022
+lammps_version: 29 Aug 2024
+tags: generated
```
What is the reason for re-generating this yaml file? Can't we just keep the original?
Also note the "generated" tag that needs to be removed.

alphataubio replied:
generated tag removed. i had regenerated the test to current stable version because i was working with it

akohlmey reviewed (changes requested):
The fix recenter changes still need to be remove.
What is there *cannot* be correct. 

We'll have to look into testing and fixing this some other time.
You'll have to stash away the force-style test yaml files as well for the time being.

alphataubio committed:
Update fix-timestep-wall_region_morse.yaml

alphataubio committed:
Update fix-timestep-wall_region_lj1043.yaml

alphataubio committed:
Update fix-timestep-wall_region_lj126.yaml

alphataubio committed:
Update fix-timestep-wall_region_harmonic.yaml

alphataubio committed:
Update fix-timestep-wall_lj93_const.yaml

alphataubio committed:
Update atom_vec_template.cpp

alphataubio committed:
Update test_atom_styles.cpp

alphataubio committed:
Update fix_recenter.cpp

alphataubio committed:
fix recenter test skipped for respa

alphataubio commented:
> The fix recenter changes still need to be remove. What is there _cannot_ be correct.
> 
> We'll have to look into testing and fixing this some other time. You'll have to stash away the force-style test yaml files as well for the time being.

fix recenter test skipped for respa

stanmoore1 committed:
Merge branch 'develop' of github.com:lammps/lammps into alphataubio-kokkos-fixes

stanmoore1 commented:
@alphataubio for the new Kokkos styles, can you please add the boiler plate note as in other styles (if it isn't in there already)? It goes near the bottom. Thanks

```

----------

.. include:: accel_styles.rst

----------

```

stanmoore1 commented:
I updated this PR to latest develop and it failed a compilation test with a weird error:
```
h5cc -I include -D_DEFAULT_SOURCE -O2 -DH5_NO_DEPRECATED_SYMBOLS  -Wall -fPIC -c src/ch5md.c -o build/ch5md.o
In file included from /usr/include/hdf5/serial/H5public.h:32,
                 from /usr/include/hdf5/serial/hdf5.h:22,
                 from src/ch5md.c:9:
/usr/include/hdf5/serial/H5version.h:39:4: error: #error "Can't choose old API versions when deprecated APIs are disabled"
   39 |   #error "Can't choose old API versions when deprecated APIs are disabled"
      |    ^~~~~
make: *** [Makefile.h5cc:21: build/ch5md.o] Error 1
make: Leaving directory '/data/workspace/dev/pull_requests/ubuntu/legacy_serial_openmp_smallsmall_static/build/lammps/lib/h5md'
++ exit 1
```

akohlmey commented:
> I updated this PR to latest develop and it failed a compilation test with a weird error:

This must be due to me updating the ubuntu container images on the Jenkins cluster from ubuntu 20.04LTS to ubuntu 22.04LTS so we can merge with PR #4344 

Update:
@stanmoore1 I am trying to fix this with https://github.com/lammps/lammps/pull/4348/commits/b2e8648e6c8f2ab46e99fdb1e0a31df73548567f
Check out the build status on PR #4344

alphataubio commented:
> > I updated this PR to latest develop and it failed a compilation test with a weird error:
> 
> This must be due to me updating the ubuntu container images on the Jenkins cluster from ubuntu 20.04LTS to ubuntu 22.04LTS so we can merge with PR #4344

i see a different compile error on jenkins

> I updated this PR to latest develop and it failed a compilation test with a weird error:
> 
> ```
> h5cc -I include -D_DEFAULT_SOURCE -O2 -DH5_NO_DEPRECATED_SYMBOLS  -Wall -fPIC -c src/ch5md.c -o build/ch5md.o
> In file included from /usr/include/hdf5/serial/H5public.h:32,
>                  from /usr/include/hdf5/serial/hdf5.h:22,
>                  from src/ch5md.c:9:
> /usr/include/hdf5/serial/H5version.h:39:4: error: #error "Can't choose old API versions when deprecated APIs are disabled"
>    39 |   #error "Can't choose old API versions when deprecated APIs are disabled"
>       |    ^~~~~
> make: *** [Makefile.h5cc:21: build/ch5md.o] Error 1
> make: Leaving directory '/data/workspace/dev/pull_requests/ubuntu/legacy_serial_openmp_smallsmall_static/build/lammps/lib/h5md'
> ++ exit 1
> ```

it works for me on macos, and rhel8 power9/v100. seems like a problem with hdf5 on your local machine. i know nothing about hdf5.

i see a different compile error on automated commit testing, it doesnt happen to me either. likely related to @akohlmey comment below.

>g++ -std=c++11 -O3 -g -fPIC -Wall  -c poemstreenode.cpp
In file included from poemstreenode.cpp:18:
poemstreenode.h: In constructor 'TreeNode::TreeNode(const int&, TreeNode*, TreeNode*, int)':
poemstreenode.h:36:13: warning: 'TreeNode::data' will be initialized after [-Wreorder]
   36 |         int data;
      |             ^~~~
poemstreenode.h:32:19: warning:   'TreeNode* TreeNode::left' [-Wreorder]
   32 |         TreeNode *left;
      |                   ^~~~
poemstreenode.cpp:22:1: warning:   when initialized here [-Wreorder]
   22 | TreeNode::TreeNode (const int & item, TreeNode *lptr,TreeNode *rptr,
      | ^~~~~~~~

akohlmey commented:
> it works for me on macos, and rhel8 power9/v100. seems like a problem with hdf5 on your local machine. i know nothing about hdf5.

It is not the local machine, but rather the Jenkins server pool at Temple. I just updated some containers with newer OS versions. The HDF5 version must have been updated which the H5MD has not been adapted. It should not be a problem with CMake builds, since it does not set the compiler define to not use deprecated APIs. Only the GNU make build does that.

> i see a different compile error on automated commit testing

These are warnings and have been around a long time. The reorder one is mostly harmless. There are others more worrisome. This is why the POEMS package will be flagged for removal after the next stable release.

stanmoore1 commented:
Internal Kokkos regression tests passed.

alphataubio committed:
Update fix_cmap.rst

alphataubio committed:
Update fix_nve_limit.rst

alphataubio committed:
Update fix_recenter.rst

alphataubio committed:
Update fix_wall_region.rst

alphataubio commented:
> @alphataubio for the new Kokkos styles, can you please add the boiler plate note as in other styles (if it isn't in there already)? It goes near the bottom. Thanks
> 
> ```
> 
> ----------
> 
> .. include:: accel_styles.rst
> 
> ----------
> ```

done

stanmoore1 committed:
Merge branch 'develop' of github.com:lammps/lammps into alphataubio-kokkos-fixes

stanmoore1 commented on src/KOKKOS/fix_cmap_kokkos.cpp:
```diff
@@ -0,0 +1,1046 @@
+{
+  kokkosable = 1;
+
+  // FIXME: test/bugfix pack_exchange_kokkos() and unpack_exchange_kokkos()
```
Please address/remove FIXME comment

alphataubio replied:
done

stanmoore1 commented on src/KOKKOS/fix_cmap_kokkos.cpp:
```diff
@@ -0,0 +1,1046 @@
+  r43 = sqrt(vb43x*vb43x + vb43y*vb43y + vb43z*vb43z);
+  a2sq = a2x*a2x + a2y*a2y + a2z*a2z;
+  b2sq = b2x*b2x + b2y*b2y + b2z*b2z;
+  //if (a1sq<0.0001 || b1sq<0.0001 || a2sq<0.0001 || b2sq<0.0001)
```
Please remove debugging comment

alphataubio replied:
done. that was alrady there in the serial class copied over

stanmoore1 commented on src/KOKKOS/fix_cmap_kokkos.cpp:
```diff
@@ -0,0 +1,1046 @@
+
+      double ecmapKK = 0.0;
+
+// FIXME: needed for compute_scalar()
```
FIXME

alphataubio replied:
done

alphataubio replied:
done

stanmoore1 commented on src/KOKKOS/fix_cmap_kokkos.cpp:
```diff
@@ -0,0 +1,1046 @@
+
+      // tally energy and/or virial
+
+/*
```
Please remove unused code block or justify why it needs to stay in

alphataubio replied:
compute_scalar() bugfix done

stanmoore1 commented on src/KOKKOS/fix_cmap_kokkos.cpp:
```diff
@@ -0,0 +1,1046 @@
+/* ----------------------------------------------------------------------
+   (KOKKOS) pack values in local atom-based array for exchange
+------------------------------------------------------------------------- */
+
```
Remove

alphataubio replied:
done

alphataubio replied:
done

stanmoore1 commented on src/KOKKOS/fix_cmap_kokkos.cpp:
```diff
@@ -0,0 +1,1046 @@
+------------------------------------------------------------------------- */
+
+/*
+int FixCMAP::unpack_exchange(int nlocal, double *buf)
```
Remove

stanmoore1 commented on src/KOKKOS/fix_nve_limit_kokkos.cpp:
```diff
@@ -0,0 +1,208 @@
+{
+  kokkosable = 1;
+
+  //FIXME: unit test fails when i turn this on
```
FIXME: you would need to add a "fused" integrate method, for example see https://github.com/lammps/lammps/blob/develop/src/KOKKOS/fix_nve_kokkos.cpp#L166.

alphataubio replied:
done

stanmoore1 commented on src/KOKKOS/fix_recenter_kokkos.cpp:
```diff
@@ -0,0 +1,141 @@
+  atomKK->sync(ExecutionSpaceFromDevice<LMPHostType>::space,X_MASK);
+  Few<double, 3> tmpxcm;
+  group->xcm(igroup,masstotal,&tmpxcm[0]);
+  const Few<double, 3> xcm(tmpxcm);
```
Is Few needed here? Seems like this is all on the host so could use a regular array instead.

alphataubio replied:
i copied that code chunk from somewhere else (fix_momentum_kokkos.cpp). fix_momentum_kokkos was done by Dan Ibanez (SNL). do you know what he means: "/* this is needed because Group is not Kokkos-aware ! */" ? 

there's no point in doing recenter/kk if it has to sync positions back to host at every step to calculate group center of mass.

alphataubio replied:
i added GroupKokkos but i cant put groupKK in pointers.h and lammps.h because GroupKokkos needs to be templated to do its work with parallel_reduce(). FixRecenterKokkos has an instance variable groupKK instead.

done

stanmoore1 commented on src/KOKKOS/fix_recenter_kokkos.h:
```diff
@@ -0,0 +1,46 @@
+    void initial_integrate(int) override;
+
+  private:
+    //typename ArrayTypes<DeviceType>::t_x_array d_x;
```
Remove if not needed

alphataubio replied:
done. what ive been doing lately is to have an auto local variable instead of an instance variable for device views, this way it avoids the problem with lambda captures of *this

stanmoore1 commented on src/KOKKOS/region_sphere_kokkos.cpp:
```diff
@@ -0,0 +1,173 @@
+  auto l_groupbit = groupbit_in;
+
+  copymode = 1;
+
```
This is correct use, can remove FIXME comment.

alphataubio replied:
i removed the FIXME but left comment to document why useKOKKOS_CLASS_LAMBDA instead of KOKKOS_LAMBDA

stanmoore1 commented on src/fix_wall_region.h:
```diff
@@ -38,12 +38,12 @@ class FixWallRegion : public Fix {
   double alpha;
   int eflag;
-  double ewall[4], ewall_all[4];
+  double *ewall, ewall_all[4]; // need ewall double*, not double[] for kokkos dual view
```
This isn't really necessary, if you copy the elements manually. @akohlmey do you care?

akohlmey replied:
> if you copy the elements manually. @akohlmey do you care?

If the change can be kept local within the KOKKOS version, that is preferred.
For consistency, either both arrays should be allocated dynamically or  with fixed size in the class.
For 4 elements a manual copy is still acceptable IMO. If I remember correctly, it is done even for the virial in places with 6 elements.

alphataubio replied:
done

stanmoore1 reviewed (changes requested):
@alphataubio thanks again for working on this, it mostly looks good, I had a few change requests mostly related to cleaning up comments.

alphataubio committed:
Update fix_cmap_kokkos.cpp

alphataubio committed:
Update fix_nve_limit_kokkos.cpp

alphataubio committed:
Update fix_recenter_kokkos.h

alphataubio committed:
Update region_sphere_kokkos.cpp

alphataubio committed:
Update fix_wall_region.h

alphataubio committed:
Update fix_wall_region.cpp

alphataubio committed:
Merge branch 'alphataubio-kokkos-fixes' of https://github.com/alphataubio/lammps-alphataubio into alphataubio-kokkos-fixes

alphataubio committed:
oops

alphataubio committed:
add GroupKokkos

alphataubio committed:
remove FIXME

alphataubio committed:
relax epsilon for kokkos_omp to pass

alphataubio committed:
Update fix_cmap.cpp

alphataubio committed:
compute_scalar() bugfix

akohlmey committed:
Merge branch 'develop' into alphataubio-kokkos-fixes

akohlmey committed:
whitespace

akohlmey committed:
revert change that breaks backward compatibility

akohlmey commented:
Compilation with Kokkos/HIP currently fails with:
```
/home/akohlmey/compile/lammps/src/KOKKOS/fix_wall_region_kokkos.cpp:181:23: warning: variable 'engKK' set but not used [-Wunused-but-set-variable]
  181 |       double fwallKK, engKK;
      |                       ^
/home/akohlmey/compile/lammps/src/KOKKOS/fix_wall_region_kokkos.cpp:173:21: error: reference to __host__ function 'surface' in __host__ __device__ function
  173 |     int n = region->surface(d_x(i,0), d_x(i,1), d_x(i,2), cutoff);
      |                     ^
/home/akohlmey/compile/lammps/src/KOKKOS/fix_wall_region_kokkos.cpp:380:16: note: in instantiation of member function 'LAMMPS_NS::FixWallRegionKokkos<Kokkos::HIP>::wall_particle' requested here
  380 | template class FixWallRegionKokkos<LMPDeviceType>;
      |                ^
/home/akohlmey/compile/lammps/src/region.h:82:7: note: 'surface' declared here
   82 |   int surface(double, double, double, double);
      |       ^
```

alphataubio committed:
port surface() to KOKKOS

alphataubio committed:
cleanup

alphataubio committed:
Update fix_wall_region_kokkos.cpp

alphataubio committed:
Update region_block_kokkos.cpp

alphataubio committed:
Update region_sphere.cpp

alphataubio committed:
Update region.cpp

alphataubio committed:
Create fix-timestep-wall_region_sphere.yaml

alphataubio committed:
add d_contact

alphataubio committed:
add two examples for regression testing

alphataubio committed:
Update region_sphere_kokkos.cpp

alphataubio committed:
Update fix_wall_region_kokkos.cpp

alphataubio committed:
remove "tags: generated"

alphataubio committed:
add_contact()

alphataubio committed:
Update in.wall.sphere

alphataubio committed:
surface_kokkos()

alphataubio committed:
dynamic_cast

alphataubio committed:
Update region_block_kokkos.h

alphataubio committed:
Update region_sphere_kokkos.h

alphataubio committed:
_kokkos

alphataubio committed:
oops

alphataubio committed:
Update fix_recenter_kokkos.cpp

alphataubio committed:
Update fix_recenter_kokkos.cpp

alphataubio committed:
functor templated on RegBlockKokkos and RegSphereKokkos

alphataubio committed:
cleanup

alphataubio committed:
remove warning

alphataubio committed:
KOKKOS_FUNCTION

alphataubio committed:
forward_transform() and find_closest_point()

alphataubio committed:
point_on_line_segment()

alphataubio committed:
match_kokkos()

alphataubio committed:
inside_face()

alphataubio committed:
Update fix_wall_region_kokkos.cpp

alphataubio committed:
sub3() and dot3()

alphataubio committed:
Update region_block_kokkos.cpp

alphataubio committed:
refactor

alphataubio committed:
all device function in headers to avoid device relocatable code with nvcc

alphataubio committed:
-diag_suppress (single dash)

alphataubio committed:
Update CMakeLists.txt

alphataubio committed:
fix nvcc option

alphataubio committed:
add tests with region sphere

alphataubio committed:
Update fix_setforce_kokkos.cpp

alphataubio committed:
Update fix_efield_kokkos.cpp

alphataubio committed:
cleanup

alphataubio committed:
partial revert of refactoring that broke efield and setforce

akohlmey committed:
silence compiler warnings

akohlmey reviewed (approved):
@stanmoore1 This looks good enough to merge to me.

alphataubio committed:
fixed identation

alphataubio committed:
Merge branch 'alphataubio-kokkos-fixes' of https://github.com/alphataubio/lammps-alphataubio into alphataubio-kokkos-fixes

akohlmey committed:
Merge branch 'develop' into alphataubio-kokkos-fixes

akohlmey commented:
@stanmoore1 this is waiting on you now

alphataubio committed:
fix warning(s)

alphataubio committed:
cuda bugfix

alphataubio committed:
cuda bugfix again

alphataubio committed:
Merge branch 'develop' into alphataubio-kokkos-fixes

stanmoore1 committed:
Merge branch 'develop' of github.com:lammps/lammps into alphataubio-kokkos-fixes

stanmoore1 commented:
Running this through internal Kokkos regression testing again.

stanmoore1 committed:
Update GNU Make for new files

stanmoore1 committed:
Fix compile issue with gcc

stanmoore1 commented:
@alphataubio @weinbe2  I absolutely cannot get this to work with `-diag_suppress 128`. I think it has to do with `nvcc_wrapper` only taking 1 argument from `-Xcudafe`.

Also it is really confusing because `nvcc --help` and the output warning message both say to use `diag-suppress` (dash) but it only works with `diag_suppress` (underscore).

I'm using CUDA 12.4, @alphataubio do you remember which Cuda version failed for you?

alphataubio commented:
> I'm using CUDA 12.4, @alphataubio do you remember which Cuda version failed for you?

i tried both cuda/11.8.0 and cuda 12.2, it didnt work with either so i had to manualy remove `diag_suppress` in my cmakelist.txt file when i build on the cluster(s).

it's a problem with a flood of fmt::format warnings, i wish we could solve that @akohlmey ?

stanmoore1 committed:
Prevent deduplication of -Xcudafe flag

masterleinad commented on cmake/CMakeLists.txt:
```diff
@@ -141,7 +141,7 @@ endif()
 # silence nvcc warnings
 if((PKG_KOKKOS) AND (Kokkos_ENABLE_CUDA) AND NOT (CMAKE_CXX_COMPILER_ID STREQUAL "Clang"))
-  set(CMAKE_TUNE_DEFAULT "${CMAKE_TUNE_DEFAULT} -Xcudafe --diag_suppress=unrecognized_pragma -Xcudafe --diag_suppress=128")
+  set(CMAKE_TUNE_DEFAULT "${CMAKE_TUNE_DEFAULT} -Xcudafe --diag_suppress=unrecognized_pragma SHELL:-Xcudafe --diag_suppress=128")
```
I'm surprised that you don't need to group the arguments clearer, like
```suggestion
  set(CMAKE_TUNE_DEFAULT "${CMAKE_TUNE_DEFAULT}" "SHELL:-Xcudafe --diag_suppress=unrecognized_pragma" "SHELL:-Xcudafe --diag_suppress=128")
```

stanmoore1 replied:
@masterleinad thanks for the suggestion. I realized both `-Xcudafe` flags can be combined into one which simplifies everything: `-Xcudafe --diag_suppress=unrecognized_pragma,--diag_suppress=128` and removes the need for the `SHELL:` prefix.

stanmoore1 committed:
Combine into a single -Xcudafe flag to simplify everything

stanmoore1 commented:
@alphataubio I think I have a good solution now, can you test?

stanmoore1 commented:
No new regression test failures.

alphataubio commented:
> @alphataubio I think I have a good solution now, can you test?

yes i just did and it works now.

alphataubio commented:
> No new regression test failures.

maybe you should add `examples/wall/in.wall.block` and `examples/wall/in.wall.sphere` to KOKKOS regression tests for RegBlockKokkos and RegSphereKokkos ?

stanmoore1 committed:
whitespace

stanmoore1 commented on src/KOKKOS/fix_cmap_kokkos.cpp:
```diff
@@ -0,0 +1,925 @@
+  auto d_buf = typename ArrayTypes<DeviceType>::t_xfloat_1d_um(
+    k_buf.template view<DeviceType>().data(),
+    k_buf.extent(0)*k_buf.extent(1));
+  //d_copylist = k_copylist.view<DeviceType>();
```
@alphataubio I think `copylist` would be needed for pack exchange, are you sure this is correct? For example see https://github.com/lammps/lammps/blob/develop/src/KOKKOS/fix_shake_kokkos.cpp#L1601-L1630.

alphataubio replied:
i made the changes for `d_copylist` in latest commit https://github.com/lammps/lammps/pull/4296/commits/4a64b3d9ced759f7c8c189bab6bea3af6036443f, but i dont know how to test `d_copylist` is working properly ?

stanmoore1 replied:
>i dont know how to test d_copylist is working properly ?

If you are running on CPU, you need to set `-pk kokkos comm device` and then test with multiple MPI ranks (exchange isn't called on a single rank).

alphataubio replied:
done. tested on openmp and cuda

alphataubio committed:
add d_exchange_sendlist to pack_exchange_kokkos()

alphataubio committed:
Merge branch 'alphataubio-kokkos-fixes' of https://github.com/alphataubio/lammps-alphataubio into alphataubio-kokkos-fixes

alphataubio committed:
Update fix_cmap_kokkos.cpp

alphataubio committed:
Merge branch 'develop' into alphataubio-kokkos-fixes

alphataubio committed:
Update fix_cmap_kokkos.cpp

alphataubio committed:
Merge branch 'alphataubio-kokkos-fixes' of https://github.com/alphataubio/lammps-alphataubio into alphataubio-kokkos-fixes

alphataubio committed:
printf debugging

alphataubio committed:
fix typos

alphataubio committed:
add vcm() and angmom()

alphataubio committed:
recommit changes from @stanmoore1

alphataubio committed:
make group kokkos aware

alphataubio committed:
cuda sync

alphataubio committed:
Update fix_cmap_kokkos.cpp

alphataubio committed:
Update fix_cmap_kokkos.cpp

alphataubio committed:
Update fix_cmap_kokkos.cpp

alphataubio committed:
Update fix_cmap_kokkos.cpp

alphataubio committed:
Update fix_cmap_kokkos.cpp

alphataubio committed:
typo

alphataubio committed:
Update fix_cmap_kokkos.cpp

alphataubio committed:
Update fix_cmap_kokkos.cpp

alphataubio committed:
Update fix_cmap_kokkos.cpp

alphataubio committed:
Update fix_cmap_kokkos.cpp

alphataubio committed:
add inertia()

alphataubio committed:
move d_ubuf to kokkos_type.h

https://matsci.org/t/a-few-kokkos-development-questions/56598/8 (Q11)
"More correct would be to use a union with ubuf as is done in the atom_vec styles, so that no precision is lost for converting a huge 64-bit integer to a double."

alphataubio committed:
this might work on cuda

alphataubio committed:
fix cuda warning

alphataubio committed:
cleanup

alphataubio committed:
Merge branch 'develop' into alphataubio-kokkos-fixes

alphataubio committed:
whitespace

alphataubio committed:
Merge branch 'alphataubio-kokkos-fixes' of https://github.com/alphataubio/lammps-alphataubio into alphataubio-kokkos-fixes

stanmoore1 commented:
Running the internal Kokkos regression tests again.

alphataubio committed:
auto variables for lambda capture

stanmoore1 commented:
Kokkos regression tests passed. I'm going to run a few of my own tests, then this should be ready to merge.

alphataubio committed:
wrap host functions with sync/modify for kokkos

alphataubio committed:
atomics

alphataubio committed:
Update fix_colvars_kokkos.cpp

alphataubio committed:
EMPTY_MASK

stanmoore1 committed:
Fix bug when a proc has no atoms/neighbors (existing issue)

stanmoore1 committed:
Fix another (already existing) bug when a proc has no atoms

stanmoore1 commented on src/KOKKOS/fix_colvars_kokkos.cpp:
```diff
@@ -0,0 +1,51 @@
+
+void FixColvarsKokkos::post_force(int vflag)
+{
+  atomKK->sync(Host,X_MASK|F_MASK);
```
@alphataubio this also needs to sync `tag` and `image`.

alphataubio replied:
done

stanmoore1 commented on src/KOKKOS/fix_colvars_kokkos.cpp:
```diff
@@ -0,0 +1,51 @@
+void FixColvarsKokkos::end_of_step()
+{
+  if (store_forces) {
+    atomKK->sync(Host,F_MASK);
```
This needs `tag`

alphataubio replied:
done

stanmoore1 committed:
Update GNU Make for new file

stanmoore1 reviewed (changes requested):
@alphataubio I finished my testing, it all appears to be working for my tests. However I see one issue in `fix_colvars_kokkos`

stanmoore1 committed:
Merge branch 'develop' of https://github.com/lammps/lammps into alphataubio-kokkos-fixes

alphataubio committed:
Update fix_colvars_kokkos.cpp

alphataubio committed:
Merge branch 'alphataubio-kokkos-fixes' of https://github.com/alphataubio/lammps-alphataubio into alphataubio-kokkos-fixes

alphataubio committed:
Merge branch 'develop' into alphataubio-kokkos-fixes

stanmoore1 committed:
Merge branch 'alphataubio-kokkos-fixes' of github.com:alphataubio/lammps-alphataubio into alphataubio-kokkos-fixes

stanmoore1 commented:
I'm seeing lost atoms with `fix recenter/kk` in a different input, need to debug.

alphataubio committed:
add syncs

alphataubio committed:
oops

stanmoore1 committed:
Merge branch 'alphataubio-kokkos-fixes' of github.com:alphataubio/lammps-alphataubio into alphataubio-kokkos-fixes

stanmoore1 commented:
@alphataubio your latest changes fixed the lost atoms issue I saw, nice work. I'm running a few more tests.

stanmoore1 committed:
Small change to simplify code

stanmoore1 reviewed (approved):
I approve

rarensu commented:
Broken for sycl
```
lammps/src/KOKKOS/fix_cmap_kokkos.cpp:886:38: error: SYCL kernel cannot use a non-const static data variable
  886 |       for (k = 0; k < 16; k++) xx += wt[in][k]*x[k]
```

alphataubio commented:
@rarensu would removing the `static` (or replacing with `const static`) on line 853 fix this problem ? 

sorry i dont access to anything SYCL (or HIP) to test.

masterleinad commented:
Adding `const` should do the trick, see https://godbolt.org/z/YxM4Wf4q1.

alphataubio commented:
@akohlmey please add this to your next small changes PR.

You're the cpp compiler expert do you think `const static` is the correct fix for this ?

akohlmey commented:
> @akohlmey please add this to your next small changes PR.
> 
> You're the cpp compiler expert do you think `const static` is the correct fix for this ?

Way ahead of you. Check out: https://github.com/lammps/lammps/pull/4387/commits/e200d557ec94cc698d6d95d0290aff6d6202420c
Which was part of the last pull request for the new release.

`const static` inside a function makes next to no sense. Just const will do. In fact, I tested it on the "godbolt" page with the same compiler setting to compile with SYCL and there was not difference in the assembly code between static const and plain const. Using static const makes more sense *outside* a function. But for that we now use static constexpr whereever possible to explicitly make it a compile time constant.
